diff --git a/src/backend/access/common/reloptions.c b/src/backend/access/common/reloptions.c
index c6aa6ab11c..766e611b11 100644
--- a/src/backend/access/common/reloptions.c
+++ b/src/backend/access/common/reloptions.c
@@ -566,6 +566,19 @@ static relopt_enum enumRelOpts[] =
 
 static relopt_string stringRelOpts[] =
 {
+	{
+		{
+			"delta_apply_function",
+			"Function called to perform delta conflict avoidance",
+			RELOPT_KIND_ATTRIBUTE,
+			ShareUpdateExclusiveLock
+		},
+		-1,
+		true,
+		NULL,
+		NULL,
+		NULL
+	},
 	/* list terminator */
 	{{NULL}}
 };
@@ -2092,6 +2105,7 @@ attribute_reloptions(Datum reloptions, bool validate)
 		{"n_distinct", RELOPT_TYPE_REAL, offsetof(AttributeOpts, n_distinct)},
 		{"n_distinct_inherited", RELOPT_TYPE_REAL, offsetof(AttributeOpts, n_distinct_inherited)},
 		{"log_old_value", RELOPT_TYPE_BOOL, offsetof(AttributeOpts, log_old_value)},
+		{"delta_apply_function", RELOPT_TYPE_STRING, offsetof(AttributeOpts, delta_apply_function)},
 		{"invisible", RELOPT_TYPE_BOOL, offsetof(AttributeOpts, invisible)}
 	};
 
diff --git a/src/include/utils/attoptcache.h b/src/include/utils/attoptcache.h
index 43e25e9b20..05fc714109 100644
--- a/src/include/utils/attoptcache.h
+++ b/src/include/utils/attoptcache.h
@@ -22,6 +22,7 @@ typedef struct AttributeOpts
 	float8		n_distinct;
 	float8		n_distinct_inherited;
 	bool		log_old_value;
+	int			delta_apply_function;
 	bool		invisible;
 } AttributeOpts;
 
