diff --git a/src/backend/commands/publicationcmds.c b/src/backend/commands/publicationcmds.c
index 803c26ab216..a60f9f1d0d8 100644
--- a/src/backend/commands/publicationcmds.c
+++ b/src/backend/commands/publicationcmds.c
@@ -683,7 +683,7 @@ check_simple_rowfilter_expr_walker(Node *node, ParseState *pstate)
  *
  * See check_simple_rowfilter_expr_walker for details.
  */
-static bool
+bool
 check_simple_rowfilter_expr(Node *node, ParseState *pstate)
 {
 	return check_simple_rowfilter_expr_walker(node, pstate);
diff --git a/src/include/commands/publicationcmds.h b/src/include/commands/publicationcmds.h
index f90cf1ef896..c7ce64dc7f2 100644
--- a/src/include/commands/publicationcmds.h
+++ b/src/include/commands/publicationcmds.h
@@ -39,5 +39,7 @@ extern bool pub_contains_invalid_column(Oid pubid, Relation relation,
 										bool *invalid_column_list,
 										bool *invalid_gen_col);
 extern void InvalidatePubRelSyncCache(Oid pubid, bool puballtables);
+extern bool check_simple_rowfilter_expr(Node *node, ParseState *pstate);
+
 
 #endif							/* PUBLICATIONCMDS_H */
