diff --git a/src/backend/executor/execReplication.c b/src/backend/executor/execReplication.c
index f262e7a66f7..0d24b056216 100644
--- a/src/backend/executor/execReplication.c
+++ b/src/backend/executor/execReplication.c
@@ -142,7 +142,7 @@ should_refetch_tuple(TM_Result res, TM_FailureData *tmfd)
 		case TM_Updated:
 			/* XXX: Improve handling here */
 			if (ItemPointerIndicatesMovedPartitions(&tmfd->ctid))
-				ereport(LOG,
+				ereport(DEBUG1,
 						(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
 						 errmsg("tuple to be locked was already moved to another partition due to concurrent update, retrying")));
 			else
@@ -153,7 +153,7 @@ should_refetch_tuple(TM_Result res, TM_FailureData *tmfd)
 			break;
 		case TM_Deleted:
 			/* XXX: Improve handling here */
-			ereport(LOG,
+			ereport(DEBUG1,
 					(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
 					 errmsg("concurrent delete, retrying")));
 			refetch = true;
